src/app_unit.rs

```diff
@@ -127,45 +141,72 @@ impl AddAssign for Au {
     #[inline]
     fn add_assign(&mut self, other: Au) {
         *self = *self + other;
+        self.clamp_self();
     }
 }
 
 impl SubAssign for Au {
     #[inline]
     fn sub_assign(&mut self, other: Au) {
         *self = *self - other;
+        self.clamp_self();
     }
 }
 
 impl MulAssign<i32> for Au {
     #[inline]
     fn mul_assign(&mut self, other: i32) {
         *self = *self * other;
+        self.clamp_self();
     }
 }
 
 impl DivAssign<i32> for Au {
     #[inline]
     fn div_assign(&mut self, other: i32) {
         *self = *self / other;
+        self.clamp_self();
     }
 }
 
 impl Au {
     /// FIXME(pcwalton): Workaround for lack of cross crate inlining of newtype structs!
     #[inline]
     pub fn new(value: i32) -> Au {
-        Au(value)
+        Au(value).clamp()
+    }
+
+    #[inline]
+    fn clamp(self) -> Self {
+        if self.0 > MAX_AU.0 {
+            MAX_AU
+        } else if self.0 < MIN_AU.0 {
+            MIN_AU
+        } else {
+            self
+        }
+    }
+
+    #[inline]
+    fn clamp_self(&mut self) {
+        *self = self.clamp()
     }
 
     #[inline]
     pub fn scale_by(self, factor: f32) -> Au {
-        Au(((self.0 as f32) * factor).round() as i32)
+        let new_float = ((self.0 as f32) * factor).round();
+        if new_float > MAX_AU.0 as f32 {
+            MAX_AU
+        } else if new_float < MIN_AU.0 as f32 {
+            MIN_AU
+        } else {
+            Au(new_float as i32)
+        }
     }
 
     #[inline]
     pub fn from_px(px: i32) -> Au {
-        Au((px * AU_PER_PX) as i32)
+        Au(px).clamp() * AU_PER_PX
```

Comment:
This `clamp` could be removed, since the multiplication will saturate and clamp anyway.